Scikit-learn 1.6 compatibility

aazuspan opened the pull request:
Closes #78 by introducing compatibility fixes for [breaking changes in sklearn >= 1.6](https://scikit-learn.org/1.6/whats_new/v1.6.html#version-1-6-0), described below.

As far as our downstream packages, if anything relied on `feature_names_in_` or `n_features_in_` they would need to be updated, but I did a quick search and didn't see either show up in the code of `sknnr-spatial`, `synthetic-knn`, or `gee-knn-python`. 

## xfail estimator checks

Instead of marking our xfailing estimator checks in the estimator tags, they're now passed as arguments to [`parametrize_with_checks`](https://scikit-learn.org/1.6/modules/generated/sklearn.utils.estimator_checks.parametrize_with_checks.html#sklearn.utils.estimator_checks.parametrize_with_checks). I'm very happy with this change since it got a lot of testing details out of the public API, but it does mean that our tests aren't backwards compatible, so I've pinned the test environment to 1.6. It would be nice to be able to test older versions, but I couldn't think of a good solution to allow that. Might be worth a future PR, though.

## Tags

Tags are [now stored in `__sklearn_tags__`](https://scikit-learn.org/1.6/developers/develop.html#estimator-tags) using dataclasses instead of as a dictionary in `_more_tags`. This was a pretty simple switch, although I tried to re-evaluate the tags while making the switch to try to make sure they were still correct and not redundant with inherited tags. Those might be worth a closer look.

We *could* keep `_more_tags` for full backwards compatibility, but I don't think they're likely to be touched by users, so I'm not too worried about breaking that.

## `validate_data`

This changed from an estimator method to a utility function. For backwards compatibility, I wrapped that function and dispatch to the appropriate place.

## `n_features_in_` and `feature_names_in_`

This is the big change that I could really use some feedback on. A while back we decided that these attributes on the transformed estimators should correspond to the raw features that were passed in by the user, rather than the transformed features that are *actually* used to fit the estimator. In other words, if you fit a `GNNRegressor` with 20 features, it might actually be fit with 5 features in CCA space, but we would show the user the original 20 features rather than exposing the implementation.

I don't think that was the wrong choice, but it did add complexity since our inherited classes run internal checks against those features. To solve that, we *were* overriding `_check_feature_names` and `_check_n_features` to just skip the check, and indirectly deferring to the corresponding checks on the transformers, but with the switch in 1.6 to utility function checks over estimator method checks, that workaround was broken.

After spending some time thinking about this and trying to come up with new workarounds, the conclusion I reached is that we're fighting a losing battle by trying to trick `sklearn` and changing the meaning of those attributes. I'm suggesting (and implemented here) that we just fall back to the default behavior where our estimators store the transformed features that they were *actually* fit with, and their transformers store the raw features that *they* were fit with.

If we think this is still a concern for users, we could expose a new attribute `raw_features_in_` or point users to `estimator.transformer_.n_features_in_`, but my suspicion is that it would rare for users to rely on these attributes.

grovduck commented on pyproject.toml:
```diff
@@ -7,7 +7,6 @@ name = "sknnr"
 dynamic = ["version"]
 description = "Scikit-learn estimators for kNN regression methods"
 readme = "README.md"
-license = ""
```
I've had this one bite me a few times recently!

grovduck commented on src/sknnr/_base.py:
```diff
@@ -13,6 +13,22 @@
     from .transformers._base import ComponentReducerMixin
 
 
+def _validate_data(estimator, *args, **kwargs):
+    """
+    Compatibility wrapper around sklearn's _validate_data function.
+
+    scikit-learn >= 1.6.0 moved _validate_data from a method of BaseEstimator to a
+    public utility function. This function wraps the utility function if available,
+    and falls back to the method if not.
+    """
+    try:
+        from sklearn.utils.validation import validate_data
+    except ImportError:
+        return estimator._validate_data(*args, **kwargs)
+
+    return validate_data(estimator, *args, **kwargs)
+
+
```
Makes total sense.  Is there a standard way to leave "breadcrumbs" for ourselves when we move to `scikit-learn>=1.6.0`?  I would think we wouldn't need this function anymore at that point.

aazuspan replied:
I'm not aware of anything more authoritative than just leaving a "TODO" comment, which I'll add.

grovduck commented on src/sknnr/_base.py:
```diff
@@ -104,33 +120,9 @@ def _set_fitted_transformer(self, X, y):
     def fit(self, X, y):
         """Fit using transformed feature data."""
-        self._validate_data(X, y, force_all_finite=True, multi_output=True)
+        _validate_data(self, X, y, ensure_all_finite=True, multi_output=True)
```
Check my logic on this one.  It looks like `ensure_all_finite` keyword was introduced in 1.6, so if the user is < 1.6.0, this will use `BaseEstimator._validate_data` which doesn't have support for that keyword, correct?  Do we need to change the keyword from `ensure_all_finite` to `force_all_finite` within the wrapper function?

aazuspan replied:
Great catch! I'll fix this in the wrapper. 

This makes me think that maybe we do need a way to easily test in < 1.6.0, to avoid more issues like this in the future... I'll give that some more thought, but will probably defer for a future PR.

grovduck commented on src/sknnr/_base.py:
```diff
@@ -104,33 +120,9 @@ def _set_fitted_transformer(self, X, y):
         """Fit and store the transformer."""
         self.transformer_ = self._get_transformer().fit(X, y)
 
-    @property
-    def feature_names_in_(self):
-        return self.transformer_.feature_names_in_
-
-    @property
-    def n_features_in_(self):
-        return self.transformer_.n_features_in_
-
-    def _check_feature_names(self, X, *, reset):
-        """Override BaseEstimator._check_feature_names to prevent errors.
-
-        This check would fail during fitting because `feature_names_in_` stores original
-        names while X contains transformed names. We instead rely on the transformer to
-        check feature names and warn or raise for mismatches.
-        """
-        return
-
-    def _check_n_features(self, X, *, reset):
-        """Override BaseEstimator._check_n_features to prevent errors.
-
-        See _check_feature_names.
-        """
-        return
-
```
Thanks for the very detailed commit message on this change.  I like your suggested methods of retrieving the raw (underlying) feature counts and names.  I agree that it wasn't worth continued workarounds that would lead us to *more* complexity.

aazuspan replied:
Sounds good, thanks for giving this some thought!

grovduck reviewed (changes requested):
@aazuspan, this looked like an involved PR :wink: - one of those that I've understood what you've done, but definitely couldn't have done it myself.  I have a few small comments, a few questions, and a upcoming issue to write based on what you've done here.  Tests are passing for me as well.

aazuspan commented on tests/test_estimators.py:
```diff
@@ -26,9 +26,90 @@
+
+        row_sum_checks = [
+            "check_regressor_multioutput",
+            "check_readonly_memmap_input",
+            "check_n_features_in_after_fitting",
+            "check_f_contiguous_array_estimator",
```
Yes, and you're spot on with the redundant checks. Thanks for catching those!

aazuspan commented:
Thanks for the eagle-eyed review @grovduck! I think I fixed all the issues you caught and addressed everything, with the possible exception of the tags. Should be ready for another quick look when you get a chance.

grovduck commented on src/sknnr/_msn.py:
```diff
@@ -67,9 +67,3 @@ class MSNRegressor(YFitMixin, OrdinationKNeighborsRegressor):
-
-    def _more_tags(self):
-        return {
-            "multioutput": True,
```
@aazuspan, I think I'm leaning toward the suggested fixes you outline here for the tags and, yes, I believe the `CCorATransformer` should match the `CCATransformer`.  Even though it's technically possible to pass the 1d `y` data, I think we want to discourage users from doing that.  My only question to you is whether we want to handle that in a separate PR?  It would be quick and dirty to do here, but perhaps the *rationale* for making that change wouldn't be as clear or might get buried.

aazuspan replied:
I think you're right that changing the *content* of the tags should be kept separate from changing the format of the tags, which was the focus of this PR. With that in mind, I'm thinking that I should revert a few changes before merging this. It's a little tricky to keep track since not every tag crosswalks and some are now inherited, but I'm thinking that I should remove the `single_output=False` and `two_d_labels=True` tags from both GNN and CCA, since those weren't previously represented in `_more_tags`. 

Can you double check my logic there?

grovduck replied:
Honestly, I'm slightly confused!  I see that previously in `GNNRegressor`, we had `_more_tags` return `multioutput=True`, but that seems to be different from the new `__sklearn_tags__` defining `single_output=False`?

But I do agree with you that those seem to be new tags that weren't in the previous `_more_tags`.

(And now that I'm digging more, there is more that I don't understand.  For example, the `estimator_type` tag on all of our transformers is `None` when it looks like you can define it as `transformer`.  Then I thought for sure `TransformerMixin` would have an `estimator_type` of `transformer`, but I get this:
```
>>> from sklearn.base import TransformerMixin
>>> print(TransformerMixin().__sklearn_tags__())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\gregorma\AppData\Local\hatch\env\virtual\sknnr\BxwCyoTQ\test\Lib\site-packages\sklearn\base.py", line 800, in __sklearn_tags__
    tags = super().__sklearn_tags__()
           ^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'super' object has no attribute '__sklearn_tags__'
```
so I'm clearly pretty confused by this new tag system.)

aazuspan replied:
> I see that previously in GNNRegressor, we had _more_tags return multioutput=True, but that seems to be different from the new __sklearn_tags__ defining single_output=False?

From my reading of [the docs](https://scikit-learn.org/dev/modules/generated/sklearn.utils.TargetTags.html), I *believe* `multi_output=True` (which we now inherit, I think from `KNeighborsRegressor`) and `single_output=False` would mean that our estimators supports *only* multi-output. 

I'm less sure on the `two_d_labels` tag, which is defined as "Whether the input is a 2D labels (y)". Given that there's a corresponding `one_d_labels`, my assumption is that it's similar to (and maybe partially redundant with) `multi_output`. I suppose the distinction is that a single-output estimator could still require 2D labels, i.e. if it reduced multiple labels to a single output.

> Then I thought for sure TransformerMixin would have an estimator_type of transformer, but I get this:

Interesting, I would have thought the same thing! Looking at the [tags for `StandardScaler`](https://github.com/scikit-learn/scikit-learn/blob/fba028b07ed2b4e52dd3719dad0d990837bde28c/sklearn/preprocessing/_data.py#L1130), most are inherited from `BaseEstimator` including `estimator_type=None`. I get the sense that the tag system is still a work in progress, and really isn't widely used outside of their testing framework, so maybe that missing tag doesn't have any practical impact. I imagine it wouldn't hurt to define it ourselves, though.

> But I do agree with you that those seem to be new tags that weren't in the previous _more_tags.

Thanks for confirming! I'll go ahead and revert those.

grovduck replied:
@aazuspan, circling back to this.  I had talked about creating a separate PR to require 2-D `y` data for `MSNRegressor` and `CCorATransformer`.  I started looking into this and came across [this comment](https://github.com/lemma-osu/sknnr/pull/44#issuecomment-1688726979) where I had convinced myself that 1-D `y` data was OK following the lead of `sklearn.cross_decomposition.CCA`.  I wanted to get your opinion before proceeding.

As I understand canonical correlation analysis, it has `X (n x p)` and `Y (n x r)` matrices.  The goal is to find coefficients for `X` and `Y` that produce linear combinations `X'` and `Y'` where `X'` and `Y'` are maximally correlated.  When I was thinking about a 1-D `Y` matrix (i.e. `r=1`), my assumption was that the user was using this in place of `Y'`, i.e. they had already found the coefficients that maximized the correlation, but only on the `Y` side.  But that doesn't really make a lot of sense.  What I failed to consider is that `Y` could be a single "species" and therefore the goal would be to find the coefficients in `X` that maximized correlation with that single species.

When looking back at the original [Moeur and Stage article](https://academic.oup.com/forestscience/article/41/2/337/4626909), it seems clear to me that their intention was that `r > 1` (at least their example data has `r=11`), but I guess the point is that *technically* `CCorA` still works if `r=1`.  (I'm honestly not totally clear if `CCA` would work similarly and still need to verify this). I guess at this point I'm leaning toward leaving the code as is, allowing for 1-D `y` arrays.

Any thoughts?  (I know this is getting buried into a PR that has no relevance to this topic, so perhaps hoisting it to a new issue makes sense.)

aazuspan replied:
I'm pretty far out of my depth here, but I think I'm *mostly* following your explanation - thanks for walking me through it.

> I guess at this point I'm leaning toward leaving the code as is, allowing for 1-D y arrays.

I agree with your conclusion here. It sounds like 1) `CCorA` and `MSN` technically work with with 1D data and 2) there are valid reasons to use 1D data (your single species example), so it makes sense to me to give users that tool and let them decide if it's appropriate for their data. There's definitely a limit to that rationale where a feature turns into a footgun, but it doesn't seem like that's where this falls.

Out of curiosity, do you know how yaImpute handles this?

grovduck replied:
> Out of curiosity, do you know how yaImpute handles this?

It appears to succeed without warning when I subset the `y` data down to a single column.  So that seems like another reason to keep with the status quo.

Which gets us back to the original intent of your PR :smile:.  Do you think it makes sense to add those additional tags to `GNNRegressor` and `CCATransformer` now that I've set the fire alarm off unnessarily?!  Probably in a new PR?

aazuspan replied:
Good to know! 

Yeah, I think that's probably a good idea. I'll open a new issue just so we can consolidate some discussion and think through any other unsettled tag questions.

grovduck reviewed (approved):
@aazuspan, your changes look good to me.  I'm still waffling on where to make the changes associated with the 1-d `y` arrays for `MSN`, but if you'd prefer handling that in a separate PR, this one looks ready to me.

aazuspan commented:
@grovduck I think we're back to parity (or as close as possible) between the new `__sklearn_tags__` and the previous `_more_tags`, so I'll go ahead and merge. Thanks for reviewing!

grovduck commented:
@aazuspan, sorry to dredge up an old topic, but I ran into this while working on `RFNNRegressor`.  I think it's taken me a while to get up to speed on the ramifications of what happened in this PR, but now I have a question for you.

In our documentation for both `GNNRegressor` and `MSNRegressor`, we still show that `n_features_in_` and `feature_names_in_` are attributes of these estimators.  Running this code, we *do* get `n_features_in_`, but even when passing with a dataframe, `feature_names_in_` is no longer an attribute of these classes:
```
>>> from sknnr.datasets import load_moscow_stjoes
>>> from sknnr import GNNRegressor
>>> X, y = load_moscow_stjoes(return_X_y=True, as_frame=True)
>>> est = GNNRegressor(n_components=3).fit(X, y)
>>> est.n_features_in_
3
>>> est.feature_names_in_
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'GNNRegressor' object has no attribute 'feature_names_in_'
```
Note that, as per your design, the *transformer* retains the correct values for `n_features_in_` and `feature_names_in_`, i.e. those values from the original (raw) data:
```
>>> est.transformer_.n_features_in_
28
>>> est.transformer_.feature_names_in_
array(['EASTING', 'NORTHING', 'ELEVMEAN', 'XSLASP', 'YSLASP', 'B1MEAN',
       'B2MEAN', 'B3MEAN', 'B4MEAN', 'B5MEAN', 'B6MEAN', 'B7MEAN',
       'B8MEAN', 'B9MEAN', 'PANMEAN', 'PANSTD', 'INTMEAN', 'INTSTD',
       'INTMIN', 'INTMAX', 'HTMEAN', 'HTSTD', 'HTMIN', 'HTMAX', 'CCMEAN',
       'CCSTD', 'CCMIN', 'CCMAX'], dtype=object)
```
You had written (very clearly) in the opening thread of this PR:
> After spending some time thinking about this and trying to come up with new workarounds, the conclusion I reached is that we're fighting a losing battle by trying to trick sklearn and changing the meaning of those attributes. I'm suggesting (and implemented here) that we just fall back to the default behavior where our estimators store the transformed features that they were actually fit with, and their transformers store the raw features that they were fit with.

It's taken me way too long to figure out what is happening, but now I think I finally understand and it's all happening in `TransformedKNeighborsRegressor.fit`:
```
    def fit(self, X, y):
        """Fit using transformed feature data."""
        _validate_data(self, X=X, y=y, ensure_all_finite=True, multi_output=True)
        self._set_dataframe_index_in(X)
        self._set_fitted_transformer(X, y)

        X_transformed = self.transformer_.transform(X)
        return super().fit(X_transformed, y)
```
When we call `super().fit(X_transformed, y)` at the last line, `X_transformed` is now an array rather than a dataframe, so it doesn't have feature names anymore when passed to `KNeighborsRegressor.fit`.  At this point the estimator still has `feature_names_in_` set to the original dataframe names, but within `validate_data` with `reset=True`, it looks for feature names on `X`, doesn't find them, and deletes the `feature_names_in_` attribute from the estimator (code [here](https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/utils/validation.py#L2712-L2720)).  

So, the question is whether we care about both `n_features_in_` and `feature_names_in_` to be set on classes derived from `TransformedKNeighborsRegressor`.  If not, I think we need to simply delete the `feature_names_in_` documentation from `GNNRegressor` and `MSNRegressor`.  If so, I think we can set this attribute from `get_feature_names_out` before returning the object, i.e. `KNeighborsRegressor.fit` becomes:
```
    def fit(self, X, y):
        """Fit using transformed feature data."""
        _validate_data(self, X=X, y=y, ensure_all_finite=True, multi_output=True)
        self._set_dataframe_index_in(X)
        self._set_fitted_transformer(X, y)

        X_transformed = self.transformer_.transform(X)
        est = super().fit(X_transformed, y)
        est.feature_names_in_ = self.transformer_.get_feature_names_out()
        return est
```
I did some testing on all five of our derived estimators (which now includes RFNNRegressor) and with `X` being passed in as both a dataframe and array and it does work.  The one caveat is that `est.transformer_.feature_names_in_` will raise an `AttributeError` if run with an array, but I think that is expected.  Note that `est.feature_names_in_` is being set to whatever `get_feature_names_out` returns from the transformer, e.g. `['cca0', 'cca1', ..., 'ccan']` for `GNNRegressor` and `['x0', 'x1', ..., 'xn']` for `MahalanobisKNNRegressor` and `EuclideanKNNRegressor`.

I know you've thought a lot more about this, so if you think setting `feature_names_in_` will screw something up, I think we should just remove the documentation.

aazuspan commented:
> The upshot (and after many futile attempts to make it work correctly) is that I think I'm trying to trick sklearn again (as you advised against) and I'm OK with just taking out the documentation that feature_names_in_ is an attribute of these estimators. Leaving n_features_in_ should be fine.

Yeah, I think we arrived at the same place. When we inherit the functionality of `sklearn`, we also inherit their strict definitions for what attributes should mean, and there are too many sneaky internal checks to ever reliably store meaningful feature names, at least with our current design (I add that caveat because I think [`Pipeline`](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html) *does* manage to run a transformer and estimator and expose the original feature names, but I think that would be a major redesign to incorporate, and probably not worth the effort if all we gain is one attribute).

Good catch on the leftover documentation, though. I agree that we should remove that, and eventually document that feature names can be accessed via the transformer instead.

For `n_features_in_`, maybe we should clarify that it's the number of transformed features which may not match the number of original features?

grovduck commented:
> I add that caveat because I think [`Pipeline`](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html) _does_ manage to run a transformer and estimator and expose the original feature names, but I think that would be a major redesign to incorporate, and probably not worth the effort if all we gain is one attribute

Funny, I think we started with `Pipeline` way back in #1, but migrated away from that.  I can't remember all the nuances, but it felt like our current architecture for having transformers within the `KNeighborsRegressor` was the right way to go.

> For `n_features_in_`, maybe we should clarify that it's the number of transformed features which may not match the number of original features?

Yep, I agree with this.  I've done this in my first documentation push for #85.

aazuspan commented:
> Funny, I think we started with Pipeline way back in https://github.com/lemma-osu/sknnr/pull/1, but migrated away from that.

You're right, I'd forgotten that your original implementation was based around a `Pipeline`! Maybe you had this figured out from the start...

> Yep, I agree with this. I've done this in my first documentation push for https://github.com/lemma-osu/sknnr/pull/85.

Sweet, looks great!

grovduck commented:
> Maybe you had this figured out from the start...

Um, no :wink:
